src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds");
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (isEmpty()) {
+            head = newNode;
+            tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> currentNode = getNodeAtIndex(index);
+            newNode.prev = currentNode.prev;
+            newNode.next = currentNode;
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null && !list.isEmpty()) {
+            for (T element : list) {
+                Node<T> newNode = new Node<>(tail, element, null);
+                if (tail != null) {
+                    tail.next = newNode;
+                }
+                tail = newNode;
+                if (size == 0) {
+                    head = newNode;
+                }
+                size++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = getNodeAtIndex(index);
+        return currentNode.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = getNodeAtIndex(index);
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
```

Comment:
You can move and call this method from getNodeAtInder() instead of using it multiple times.

Reply:
yeah, thx